File: Orimath.ViewPlugins/Plugins/IMessenger.fs
namespace Orimath.Plugins
open System
open System.Windows.Input
open ApplicativeProperty


type ViewPane =
    | Main = 0
    | Top = 1
    | Bottom = 2
    | Side = 3
    | Dialog = 4


type IMessenger =
    abstract member RootEnable : IGetProp<bool>
    abstract member AddViewModel : viewModel: obj -> unit
    abstract member RemoveViewModel : viewModelType: Type -> unit
    abstract member RemoveViewModel : viewModel: obj -> unit
    abstract member RegisterView : viewPane: ViewPane * viewModelType: Type * viewType: Type -> unit
    abstract member RegisterView : viewPane: ViewPane * viewModelType: Type * viewUri: string -> unit
    abstract member SetEffectParameterViewModel<'ViewModel> : mapping: ('ViewModel -> obj) -> unit
    abstract member OpenDialog : viewModel: obj -> unit
    abstract member CloseDialog : unit -> unit
    abstract member CloseDialogCommand : ICommand
    abstract member GetEffectCommand : effect: IEffect -> ICommand
    abstract member SaveSetting : name: string * model: 'T -> unit
    abstract member LoadSetting : name: string -> 'T option
    abstract member LocalizeWord : text: string -> string
    abstract member ShowMessage : message: string -> unit

File: Orimath.ViewPlugins/Plugins/IMessenger.cs
using System;
using System.Windows;
using System.Windows.Input;

namespace Orimath.Plugins
{
    public enum ViewPane
    {
        Main = 0,
        Top = 1,
        Side = 2,
        Dialog = 3,
    }

    public interface IMessenger
    {
        void AddViewModel(object viewModel);
        void RemoveViewModel(Type viewModelType);
        void RemoveViewModel(object viewModel);
        void RegisterView(ViewPane viewPane, Type viewModelType, Type viewType);
        void RegisterView(ViewPane viewPane, Type viewModelType, string viewUri);
        void SetEffectParameterViewModel<TViewModel>(Func<TViewModel, object> mapping);
        void OpenDialog(object viewModel);
        void CloseDialog();
        ICommand CloseDialogCommand { get; }
        ICommand GetEffectCommand(IEffect effect);
    }

    public static class MessengerExtensions
    {
        public static void RegisterView<TViewModel, TView>(this IMessenger messenger, ViewPane viewPane)
            where TView : FrameworkElement
        {
            messenger.RegisterView(viewPane, typeof(TViewModel), typeof(TView));
        }
    }
}
